packages/babylon/test/fixtures/experimental/pipeline-operator/await-expr/input.js

```diff
@@ -0,0 +1,3 @@
+async function foo() {
+  return a |> await b |> f;
```

Comment:
This should result in a parse error.

Reply:
Do we want it to result in a parse error? The current pipeline plugin transforms this behavior. My approach was to not break the current plugin, which matches the current spec.

Reply:
My understanding is the pipeline operator is F# style proposal, which we follow until another proposal is presented to TC39. So this should error out, but `a |> (await b) |> f` shouldn't.

Reply:
The current spec hasn't made `|> await` normative yet: https://github.com/tc39/proposal-pipeline-operator/pull/85 – so it's not exactly the same as the F#-style, which does. I think we're holding on that PR until we settle the F#- v Hack-style debate. My goal was to make it possible to match all 3 (current spec, F#-style, & Hack-style).

Reply:
On second thought, I think this _needs_ to parse in order to implement the Hack-style pipelining, which still has a RHS for the `await` statement.

Reply:
@jridgewell: All four current proposals (Proposal 1: F-sharp Style Only with `await` special case, Proposal 2: Hack Style Only, Proposal 3: Split Mix, and Proposal 4: Smart Mix) are at equal current status, and all of them different from the current proposal at Stage 1. @littledan’s last stated his plans at https://github.com/tc39/proposal-pipeline-operator/issues/89#issuecomment-363853394, which are to present all options to TC39 as an update. They’re all on equal footing right now as long as the operator in general remains at Stage 1.

@mAAdhaTTah: I’m interested in why this parser change would be needed for Proposals 2,3,4. Doesn’t it just need `expression |> await expression` and not a bare `expression |> await`?

Reply:
@js-choi No, @jridgewell is suggesting that `expr |> await expr` should be a SyntaxError; doing so would make it impossible (I believe) to implement proposals 2-4.

Reply:
I agree that the grammar of these proposals differs from each other, and it's not the kind of relationship where one is a subset of the other. I don't know what the best way to handle this is technically within Babel, but it seems to me like Babylon will eventually need some kind of mode switch between multiple options. I think it'd be valuable to have both @js-choi and @mAAdhaTTah 's alternatives land in Babel to be usable by actual programmers without them having to check out a custom branch.

Reply:
@littledan @js-choi I think we can proceed initially w/ proposals 2-4 w/o changing babylon if we start with the placeholder being a valid identifier, e.g. using @gilbert's preference for `$_`. Once we have that implemented, we then update babylon to parse the `#` as an identifier only within a pipeline, and from there, it shouldn't be too difficult to modify the pipeline plugin to provide the placeholder as a configuration option, including using `#` (or even `?`) as a placeholder.

Reply:
@mAAdhaTTah Maybe that's a reasonable way to stage this, to implement it little by little and land incremental patches, but at some point, I think you'll need to update Babylon. I don't see the problem with that, though.

Reply:
@js-choi:

> This realization might complicate things, since Babylon doesn’t have a plugin system, right? In order for Babel to support both our Babel plugins and to let TC39 compare our proposals, Babylon needs to parse both.

I had been wrong here; I had forgotten that [Babylon does have a “plugin system”][plugins]. It’s just [not open to third parties][FAQ]. This pull request modifies the existing `pipelineOperator` plugin.

I wonder if it would be better to isolate Proposal 1 and Proposal 4 into separate Babylon plugins, both separate from the current `pipelineOperator` plugin. As @littledan says, it depends on how much their grammars have in common. Alternatively, we could both modify the existing `pipelineOperator` plugin but require a `pipelineMode` option (we could use the [`babylon.parse` options parameter][Babylon options] – can Babylon plugins themselves accept options too?)

As I said, I’m uncertain about what the precedence and associativity of Proposal 1’s `|>` is—but the biggest difference between Proposal 1 and 4 are how they handle arrow functions. `x |> $ => $ + 1` would be idiomatic with Proposal 1 but it would be an early error for Proposal 4, in which you would write `x |> # + 1` instead. This difference is what, above all, makes the two proposals mutually exclusive. (Bare `await` in `x |> await` is also forbidden by Proposal 4, but Proposal 4 could hypothetically be extended to allow it.)

A hypothetical version of Proposal 1 – one that accepts only [Proposal-4-style “simple references”][simple reference] at its `|>` RHS – could be forward compatible with Proposal 4. That is, a version of Proposal 1 whose `|>` is also right associative and also just looser than ternary conditional – and which allows `x |> functionIdentifier` and `x |> object.method` but forbids `x |> $ => $ + 1` – would be forward compatible with Proposal 4.

But this solution probably would overcomplicate both proposals’ Babylon implementations. It might be cleanest to just keep both proposals separate.

@mAAdhaTTah: As @littledan says, eventually we will need to update Babylon for Proposal 4 anyway. I feel that it might be best to deal with that now, early on.

@littledan, @mAAdhaTTah: So the three simplest options I see are:

* A. Put both Proposals 1 and 4 into the existing `pipelineOperator` plugin but require a `pipelineMode` option (either using the [`babylon.parse` options parameter][Babylon options] or a plugin-specific option).

  1. @mAAdhaTTah would modify this fork and this pull so that his changes to the existing `pipelineOperator` plugin are isolated behind a `pipelineMode: 'fSharpWithBareAwait'` flag or whatever.
  2. I would fork @mAAdhaTTah’s branch and isolate my changes behind a `pipelineMode: ’smartMix'` flag.
  3. I would eventually pull request @mAAdhaTTah’s branch with a `smartPipelines` plugin.
  4. After that, @mAAdhaTTah’s branch would in turn eventually be pulled into upstream Babylon, either through this pull request or another pull request.
  5. Our Babel plugins would both use a single Babylon plugin `pipelineOperator`, respectively using `pipelineMode: 'fSharpWithBareAwait'` and `pipelineMode: 'smartMix'`.
  6. Issue: Can Babel plugins configure Babylon using specific Babylon options?

* B. Modify the `pipelineOperator` plugin to match Proposal 1, including its bare `|> await` form. Put Proposal 4 into a separate `smartPipelines` plugin that is mutually exclusive with `pipelineOperator`.

  1. @mAAdhaTTah would keep this fork and this pull request the same, which directly modifies the existing `pipelineOperator` plugin.
  2. I would fork upstream Babylon with a `smartPipelines` plugin, which would be mutually exclusive with `pipelineOperator`. I would separately pull request upstream Babylon with `smartPipelines`.
  3. Our Babel plugins would respectively depend on Babylon `pipelineOperator` and Babylon `smartPipelines`.

* C. Isolate Proposal 1 and Proposal 4 into two mutually exclusive Babylon plugins `pipelineBareAwait` and `smartPipelines`, both of which are separate from the current `pipelineOperator` plugin.

  1. @mAAdhaTTah would modify this fork and pull request so that his changes are put in another `pipelineBareAwait` plugin.
  2. I would fork upstream Babylon with a `smartPipelines` plugin, which would be mutually exclusive with `pipelineBareAwait`. I would separately pull request upstream Babylon with `smartPipelines`.
  3. Our Babel plugins respectively depend on Babylon `pipelineOperator`+`pipelineBareAwait` and Babylon `pipelineOperator`+`smartPipeline`.

Or maybe there are other options. I’m fine with whatever, as long as I get to independently muck with the parser too.

[plugins]: https://github.com/babel/babel/blob/master/packages/babylon/README.md#plugins
[FAQ]: https://github.com/babel/babel/blob/master/packages/babylon/README.md#will-babylon-support-a-plugin-system
[simple reference]: https://jschoi.org/18/es-smart-pipelines/spec#prod-SimpleReference
[Babylon options]: https://github.com/babel/babel/blob/master/packages/babylon/README.md#options

Reply:
@js-choi I think you're overthinking it; we should be able to make babylon produce an AST compatible with all of the proposals, and the have the pipeline plugin itself be responsible for throwing errors when it encounters syntax that isn't valid for the mode it's in, e.g. in F#-mode, we'd throw if we found `|> await a`, while Hack-mode would throw on `|> await`, etc.

This might make babylon more forgiving around the pipeline operator than it should be, but I think that's ok for now, as the goal here is to get this into the wild and get feedback from developers, and we can tighten up babylon's behavior once pipeline is stage 2.

As I mentioned above, if you start with the valid-identifier version, I don't think switching that to `#`, once the changes are made in babylon, will be difficult.